src/Microsoft.AspNet.JsonPatch/JsonPatchDocument.cs

```diff
@@ -0,0 +1,210 @@
+                  null,
+                  null,
+                  Resources.FormatInvalidValueForPath(path)));
+            }
+
+            Operations.Add(new Operation("replace", checkPathResult.AdjustedPath, null, value));
+            return this;
+        }
+
+        public JsonPatchDocument Move([NotNull] string from, [NotNull] string path)
+        {
+            var checkPathResult = PathHelpers.CheckPath(path);
+            var checkFromResult = PathHelpers.CheckPath(from);
+
+            if (!checkPathResult.IsCorrectlyFormedPath)
+            {
+                LogError(new JsonPatchError(
+                   null,
+                   null,
+                   Resources.FormatInvalidValueForPath(path)));
+            }
+
+            if (!checkFromResult.IsCorrectlyFormedPath)
+            {
+                LogError(new JsonPatchError(
+                  null,
+                  null,
+                  Resources.FormatInvalidValueForPath(from)));
+            }
+
+
+            Operations.Add(new Operation("move", checkPathResult.AdjustedPath, checkFromResult.AdjustedPath));
+            return this;
+        }
+
+        public JsonPatchDocument Copy([NotNull] string from, [NotNull] string path)
+        {
+            var checkPathResult = PathHelpers.CheckPath(path);
+            var checkFromResult = PathHelpers.CheckPath(from);
+
+            if (!checkPathResult.IsCorrectlyFormedPath)
+            {
+                LogError(new JsonPatchError(
+                  null,
+                  null,
+                  Resources.FormatInvalidValueForPath(path)));
+            }
+
+            if (!checkFromResult.IsCorrectlyFormedPath)
+            {
+                LogError(new JsonPatchError(
+                  null,
+                  null,
+                  Resources.FormatInvalidValueForPath(from)));
+            }
+
+            Operations.Add(new Operation("copy", checkPathResult.AdjustedPath, checkFromResult.AdjustedPath));
+            return this;
+        }
+
+
+        public void ApplyTo([NotNull] object objectToApplyTo)           
+        {
+            ApplyTo(objectToApplyTo, new ObjectAdapter(ContractResolver, logErrorAction: null));
+        }
+
+        public void ApplyTo([NotNull] object objectToApplyTo, Action<JsonPatchError> logErrorAction)
+          
+        {
+            ApplyTo(objectToApplyTo, new ObjectAdapter(ContractResolver, logErrorAction));
+        }
+
+        public void ApplyTo([NotNull] object objectToApplyTo, [NotNull] IObjectAdapter adapter)            
+        {
+            // apply each operation in order
+            foreach (var op in Operations)
+            {
+                op.Apply(objectToApplyTo, adapter);
+            }
+        }
+
+        // LogError method is required on untyped JsonPatchDocument, as errors may 
+        // be thrown in case of invalid paths.
+
+        private void LogError([NotNull] JsonPatchError jsonPatchError)
```

Comment:
This code exists in `ObjectAdapter` too. We should put it at a separate class and use it this and the other class to avoid repetition?